File: src_python/dihed/qnmat/qnmat_1.py
import sys
import numpy as np
import cython
from typing import Self
from numpy.typing import NDArray

import crsys as crs
import qnnum as qnn
import qnvec as qnv
import qnndarray as qna
from numpy.typing import(NDArray)

# this includes nD Qnvec as a special case (shape=(n))
#class Qnmat(np.ndarray):
class Qnmat(qna.QnNdarray):
    def __new__(cls, shape:np.int64):
        return super().__new__(cls,shape)
        #return super().__new__(cls,shape,dtype=qnn.Qnnum)

    def __init__(self,shape:np.int64):
        #global n,N
        qn0=qnn.zero()
        self.N=N
        self.shape=shape
        ndim=len(shape)
        #print("ndim in Qnmat",ndim)  # for test
        if ndim==1:
            for i in range(shape[0]):
                self[i]=qn0
        
        if ndim==2:
            for i in range(shape[0]):
                for j in range(shape[1]):
                    self[i][j]=qn0
        #print("self.shape",self.shape)
        #print("self.ndim",self.ndim)
        #print("self.dtype",self.dtype)
        #printqnm("Qnmat self",self) # for test

    def __add__(ma1:Self, ma2:Self):  #  for ma1+ma2
        return add(ma1,ma2)
    
    def __iadd__(ma1:Self, ma2:Self):  #  for ma1+=ma2
        return iadd(ma1,ma2)
    
    def __isub__(ma1:Self, ma2:Self):  #  for ma1-=ma2
        return isub(ma1,ma2)
        
    def __matmul__(ma1:Self, ma2:Self):  #  for ma1@ma2
        return mul(ma1,ma2)
        #if isinstance(ma2,qna.QnNdarray):
        #    return mul(ma1,ma2)
        #if isinstance(ma2,NDArray[np.int64]):
        #    return mul_i(ma1,ma2)
    
    def set_mt(self:Self,mt:np.array): # QnNdarray*
        shape=self.shape
        for i in range(shape[0]):
            for j in range(shape[1]):
                self.mt[i][j]=qnn.copy(mt[i][j]) # copy qnnum
                
    def __copy__(a:Self):
        return copy(a)
    
def qnmat_init():
    global n,N,scly,isys
    n=crs.n
    N=crs.N
    isys=crs.isys
    if isys==3:
        scly=2.0*np.sin(np.pi/5)
    else:
        scly=1.0

def zerom(shape:np.int64) -> Qnmat:
    qnm=Qnmat(shape)
    return qnm

def zeroms(shape:np.int64) -> Qnmat:
    nm=shape[0]
    n=shape[1]
    #qnms=[Qnmat((shape[1],shape[2])) for i in range(n)]
    qnms=qna.zeros((nm),dtype=qnn.Qnmat)
    for i in range(nm):
        qnvs[i]=zerom((shape[1],shape[2]))
    return qnvs
    return qnms

File: src_python/dihed/qnmat/cpp/qnmat_1.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include "crsys.h"
#include "qnnum.h"
#include "qnvec.h"
#include "qnndarray.h"

class Qnmat : public QnNdarray {
public:
    static Qnmat* create(const std::vector<int64_t>& shape) {
        return new Qnmat(shape);
    }

    Qnmat(const std::vector<int64_t>& shape) : QnNdarray(shape) {
        Qnnum qn0 = qnn::zero();
        this->N = N;
        this->shape = shape;
        int ndim = shape.size();

        if (ndim == 1) {
            for (int64_t i = 0; i < shape[0]; ++i) {
                (*this)[i] = qn0;
            }
        }

        if (ndim == 2) {
            for (int64_t i = 0; i < shape[0]; ++i) {
                for (int64_t j = 0; j < shape[1]; ++j) {
                    (*this)[i][j] = qn0;
                }
            }
        }
    }

    Qnmat operator+(const Qnmat& ma2) const {
        return add(*this, ma2);
    }

    Qnmat& operator+=(const Qnmat& ma2) {
        return iadd(*this, ma2);
    }

    Qnmat& operator-=(const Qnmat& ma2) {
        return isub(*this, ma2);
    }

    Qnmat operator@(const Qnmat& ma2) const {
        return mul(*this, ma2);
    }

    void set_mt(const std::vector<std::vector<Qnnum>>& mt) {
        auto shape = this->shape;
        for (int64_t i = 0; i < shape[0]; ++i) {
            for (int64_t j = 0; j < shape[1]; ++j) {
                this->mt[i][j] = qnn::copy(mt[i][j]);
            }
        }
    }

    Qnmat* copy() const {
        return copy(*this);
    }
};

void qnmat_init() {
    global n, N, scly, isys;
    n = crs::n;
    N = crs::N;
    isys = crs::isys;
    if (isys == 3) {
        scly = 2.0 * std::sin(M_PI / 5);
    } else {
        scly = 1.0;
    }
}

Qnmat* zerom(const std::vector<int64_t>& shape) {
    return Qnmat::create(shape);
}

std::vector<Qnmat*> zeroms(const std::vector<int64_t>& shape) {
    int64_t nm = shape[0];
    int64_t n = shape[1];
    std::vector<Qnmat*> qnms(nm);
    for (int64_t i = 0; i < nm; ++i) {
        qnms[i] = zerom({shape[1], shape[2]});
    }
    return qnms;
}

